src/exo/LoopIR_scheduling.py

```diff
@@ -1493,6 +1504,194 @@ def map_e(self, e):
+            writes += get_writes_of_stmts(s.body)
+        elif isinstance(s, (LoopIR.WindowStmt, LoopIR.WriteConfig)):
+            raise NotImplementedError("WindowStmt and WriteConfig not supported yet")
+        elif isinstance(s, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free, LoopIR.Call)):
+            pass
+        else:
+            raise NotImplementedError(f"unknown stmt type {type(s)}")
+    return writes
+
+
+def DoLiftConstant(assign_cursor, loop_cursor):
+    orig_proc = assign_cursor.get_root()
+    assign_s = assign_cursor._node
+    loop = loop_cursor._node
+    constant = None
+    constant_src_stmt = None
+
+    for (name, typ) in get_reads_of_stmts(loop.body):
+        if assign_s.name == name and assign_s.type == typ:
+            raise SchedulingError(
+                "cannot lift constant because the buffer is read in the loop body"
+            )
+
+    def check_only_scaled_reduces(stmts):
+        nonlocal constant
+        nonlocal constant_src_stmt
+        check = True
+        for s in stmts:
+            if isinstance(s, LoopIR.Assign):
+                if s.name == assign_s.name:
+                    return False
+            elif isinstance(s, LoopIR.Reduce):
+                if s.name == assign_s.name:
+                    if (
+                        same_write_dest(orig_proc, assign_s, s)
+                        and isinstance(s.rhs, LoopIR.BinOp)
+                        and s.rhs.op == "*"
+                    ):
+                        if constant is not None:
+                            if isinstance(constant, LoopIR.Const):
+                                if (
+                                    not isinstance(s.rhs.lhs, LoopIR.Const)
+                                    or constant.val != s.rhs.lhs.val
+                                ):
+                                    return False
+                            elif isinstance(constant, LoopIR.Read):
+                                if (
+                                    not isinstance(s.rhs.lhs, LoopIR.Read)
+                                    or constant.name != s.rhs.lhs.name
+                                    or not same_index_exprs(
+                                        orig_proc,
+                                        constant.idx,
+                                        constant_src_stmt,
+                                        s.rhs.lhs.idx,
+                                        s,
+                                    )
+                                ):
+                                    return False
+                            else:
+                                raise TypeError("unknown type for self.constant")
+                        else:
+                            if isinstance(s.rhs.lhs, (LoopIR.Const, LoopIR.Read)):
+                                constant = s.rhs.lhs
+                                constant_src_stmt = s
+                            else:
+                                return False
```

Comment:
Reduce nesting by writing:

```
if not isinstance(s.rhs.lhs, (LoopIR.Const, LoopIR.Read)):
    return False

constant = s.rhs.lhs
constant_src_stmt = s
```

Continue to apply this strategy ("return asap upon failure") to make this code more straight-line.